packages/cosmic-swingset/src/helpers/sync-xfer.js

```diff
@@ -0,0 +1,160 @@
+// @ts-check
+
+import { Far } from '@endo/far';
+
+// FIXME: for debugging the buffer splitting/rejoining:
+// const DEFAULT_BUFFER_SIZE = 12;
+const DEFAULT_BUFFER_SIZE = 64 * 1024;
+export { DEFAULT_BUFFER_SIZE };
+
+// const TRANSFER_UNSPECIFIED = 0;
+const TRANSFER_CLIENT_READY = 1;
+const TRANSFER_READY_TO_READ = 2;
+const TRANSFER_READ_PARTIAL = 3;
+const TRANSFER_READ_DONE = 4;
+
+const { Fail } = assert;
+
+const debug = (..._args) => {};
+
+/**
+ * @template T
+ * @typedef {Generator<undefined, T, Uint8Array | undefined>} Decoder<T>
+ */
+
+/**
+ * @typedef {Generator<Uint8Array | undefined, Uint8Array, Uint8Array>} Encoder<T>
+ */
+
+/**
+ * @returns {Decoder<string>}
+ */
+export function* makeTextDecoder() {
+  // console.error('created decoder!');
+  const td = new TextDecoder('utf-8');
+  let s = '';
+  for (;;) {
+    const buf = yield;
+    // console.error('got buf', buf);
+    if (buf === undefined) {
+      return s;
+    }
+    s += td.decode(buf, { stream: true });
+  }
+}
+harden(makeTextDecoder);
+
+/**
+ * @param {string} input
+ * @returns {Encoder}
+ */
+export function* makeTextEncoder(input) {
+  const te = new TextEncoder();
+
+  let buf = yield;
+
+  // TODO: Maybe encode partials directly into buf, not create a fresh one.
+  buf = te.encode(input);
+  return buf;
```

Comment:
The suggestion does make sense to me, and dealing with things here would eliminate the nested loop in TransferServer `write`.
```suggestion
  for (let remaining = input, buf, read, written; ;) {
    buf = yield { buf, written };

    // With no buffer provided, encode whatever remains into a new buffer and return.
    if (buf === undefined) {
      const rest = te.encode(remaining);
      return { buf: rest, written: rest.length };
    }

    // Otherwise, write into the provided buffer and check if it's time to finish.
    ({ read, written } = te.encodeInto(remaining, buf));
    remaining = remaining.slice(read);
    if (!remaining) return { buf, written };
  }
```